### ISSUE TITLE
[ISSUE for v2]: WASAPI for Server Audio cannot be captured in other recording software during usage 

### ISSUE THREAD

**Issue by JarodMica** [w-okada/voice-changer#1305]
### Voice Changer Version

vcclient_win_std_2.0.58-alpha.zip

### Operational System

Windows 11

### GPU

RTX 4090

### CUDA Version

12.1

### Read carefully and check the options

- [ ] If you use win_cuda_torch_cuda edition, setup cuda? [see here](https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements)
- [ ] If you use win_cuda edition, setup cuda and cudnn? [see here](https://onnxruntime.ai/docs/execution-providers/CUDA-ExecutionProvider.html#requirements)
- [ ] If you use mac edition, client is not launched automatically. Use chrome to open application.?
- [X] I've tried to change the Chunk Size
- [ ] I've tried to set the Index to zero
- [X] I've read the [tutorial](https://github.com/w-okada/voice-changer/blob/master/tutorials/tutorial_rvc_en_latest.md)
- [X] I've tried to extract to another folder (or re-extract) the .zip file

### Does pre-installed model work?

YES

### Model Type

Beatrice v2

### Issue Description

Windows WASAPI in server mode for v2 of the voice changer doesn't allow for output to other programs (OBS, windows sound, etc.).

![image](https://github.com/user-attachments/assets/81d528df-5e2e-48d9-a361-893457228392)

## Example Set-up
Set-up:
OBS Mic - USB Audio Codec
VC Mic - USB Audio Codec
OBS Out - Headphones 1
VC Mic - Headphones 1

Expected behavior with WASAPI on Server Audio for OBS:
1. Click start in voice changer
2. Speak into microphone
3. Click stop
4. OBS should capture both USB Audio Codec and Headphones 1

Observed behavior:
1. Click start in voice changer
2. Speak into microphone
3. Click stop
4. OBS did not capture both USB Audio Codec and Headphones 

This is not an issue in v1 but seems to be an issue with the newer version of the voice changer for WASAPI in server audio.  I don't see this being an issue with OBS because it works properly in v1.

Additionally, other sound drivers like MME and Windows DirectSound seem work, but WASAPI has the lowest latency out of all of these so it'd be nice if this was available to work.

### Logs on console

No logs in the terminal window, but these are some parts from the vcclient logs.  Let me know if you need others parts of the log.

```
2024-08-08 13:07:47,016 - uvicorn.ac - h11_impl             - INFO - 127.0.0.1:53861 - "GET /api/voice-changer-manager/information HTTP/1.1" 200 - uvicorn\protocols\http\h11_impl.py - 477
2024-08-08 13:07:47,108 - vcclient   - voice_changer        - WARNING - start_convert_chunk_bulk called - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 247
2024-08-08 13:07:47,132 - vcclient   - local_voice_changer_ - INFO - audio input device:38, sr:48000.0 - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 95
2024-08-08 13:07:47,132 - vcclient   - local_voice_changer_ - INFO - audio output device:36, sr:48000.0 - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 96
2024-08-08 13:07:47,132 - vcclient   - local_voice_changer_ - INFO - audio monitor device:-1, sr:-1 - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 97
2024-08-08 13:07:47,133 - vcclient   - local_voice_changer_ - INFO - block size:480, 480, 0 - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 113
2024-08-08 13:07:47,190 - vcclient   - voice_changer        - INFO - resample_ratio_in:0.3333333333333333, resample_ratio_out:2.0, resample_ratio_monitor:-4.1666666666666665e-05 - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 121
2024-08-08 13:07:47,190 - vcclient   - voice_changer        - INFO -   self.vc_info.vc_input_sample_rate:16000, self.input_sample_rate:48000 - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 122
2024-08-08 13:07:47,190 - vcclient   - voice_changer        - INFO -   self.vc_info.vc_output_sample_rate:24000, self.output_sample_rate:48000 - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 123
2024-08-08 13:07:47,190 - vcclient   - voice_changer        - INFO -   self.vc_monitor_sample_rate:24000, self.monitor_sample_rate:-1 - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 124
2024-08-08 13:07:47,191 - vcclient   - voice_changer        - WARNING - data type resampled is short. padded.:(17,), shape:(160,) - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 514
2024-08-08 13:07:47,209 - vcclient   - local_voice_changer_ - INFO - Start Local Voice Changer(1way), sr:[48000.0, 48000.0], ch:[1, 2], depth:[float32, float32] - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 181
2024-08-08 13:07:51,470 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,470 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,489 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,490 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,509 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,510 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,529 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,530 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,549 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,550 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,562 - vcclient   - local_voice_changer_ - DEBUG - audio output callback ex: empty queue : - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 270
2024-08-08 13:07:51,570 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,570 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,589 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,589 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,610 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,610 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,629 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,629 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,650 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,650 - vcclient   - voice_changer        - WARNING - bulk conversion process is not started. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 269
2024-08-08 13:07:51,665 - vcclient   - local_voice_changer_ - DEBUG - audio output callback ex: empty queue : - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 270
2024-08-08 13:07:51,673 - vcclient   - local_voice_changer_ - INFO - Local Voice Changer main thread stopped - vcclient_dev\voice_changer\voice_change_manager\local_voice_changer_interface.py - 191
2024-08-08 13:07:51,955 - vcclient   - voice_changer        - DEBUG - bulk conversion process awared the stop operation. - vcclient_dev\voice_changer\voice_change_manager\voice_changer.py - 357
2024-08-08 13:07:51,955 - uvicorn.ac - h11_impl             - INFO - 127.0.0.1:53861 - "POST /api/voice-changer/operation/stop_server_device HTTP/1.1" 200 - uvicorn\protocols\http\h11_impl.py - 477
2024-08-08 13:07:51,958 - vcclient   - voice_changer_manage - INFO - Getting voice changer manager information. - vcclient_dev\voice_changer\voice_change_manager\voice_changer_manager.py - 22
```

**Comment by Kuuko-fokkusugaru** [comment]
Did you check if there are duplicates of the same outputs and inputs in the available devices list?
Your setup is a bit confusing so I didn't quite understand what were you trying to do. But I do know that certain driver access modes are exclusives which may not allow other software to access to them.

**Comment by JarodMica** [comment]
> Did you check if there are duplicates of the same outputs and inputs in the available devices list? Your setup is a bit confusing so I didn't quite understand what were you trying to do. But I do know that certain driver access modes are exclusives which may not allow other software to access to them.

There aren't any duplicates in this case, so I could rule this one out. I couldn't find a way write it out well, but maybe a video can help show what's happening:  https://youtu.be/48DXyPYwFug

I can understand drivers being exclusive, but the main issue is that I use the same settings and drivers in VC 1 and VC 2. I don't show it in the video, but 1.15.17b version of the software is capable of being recorded while using wasapi

**Comment by Kuuko-fokkusugaru** [comment]
> > Did you check if there are duplicates of the same outputs and inputs in the available devices list? Your setup is a bit confusing so I didn't quite understand what were you trying to do. But I do know that certain driver access modes are exclusives which may not allow other software to access to them.
> 
> There aren't any duplicates in this case, so I could rule this one out. I couldn't find a way write it out well, but maybe a video can help show what's happening: https://youtu.be/48DXyPYwFug
> 
> I can understand drivers being exclusive, but the main issue is that I use the same settings and drivers in VC 1 and VC 2. I don't show it in the video, but 1.5.3.17b version of the software is capable of being recorded while using wasapi

Oh I see your issue now. I never use server mode because, even when used in a different computer, I always use the local devices instead of the ones hooked to the server. But I understand your reasoning about lowering delay as much as you can if wasapi helps you on that. My suggestion was just in case since sometimes people miss little details. In this case, sadly, it didn't help so we need Okada to further look into it. There are some prior versions that should be for download too like 0.55 and 0.56 I believe. I am not sure if these may have this same issue or not, just in case that you need a quick replacement and those happens to work fine.

Btw I didn't notice until I watched the video but I have seen your videos a lot of times :D. Thank you for your hard work 💛. They are very handy to be up to date in the tech. (end of the off topic lol).

**Comment by JarodMica** [comment]
> Oh I see your issue now. I never use server mode because, even when used in a different computer, I always use the local devices instead of the ones hooked to the server. But I understand your reasoning about lowering delay as much as you can if wasapi helps you on that. My suggestion was just in case since sometimes people miss little details. In this case, sadly, it didn't help so we need Okada to further look into it. There are some prior versions that should be for download too like 0.55 and 0.56 I believe. I am not sure if these may have this same issue or not, just in case that you need a quick replacement and those happens to work fine.

Ah gothca, no worries!  We'll have to wait on okada as even if I wanted to debug the versions, it doesn't look like the source code is up-to-date as v2 is missing from the repo.  0.56 and it was still there, so I made the generalization that something changed in v2 as a whole for handling these drivers.

> Btw I didn't notice until I watched the video but I have seen your videos a lot of times :D. Thank you for your hard work 💛. They are very handy to be up to date in the tech. (end of the off topic lol).

Glad to see my content has spread wide and far :)

**Comment by w-okada** [comment]
We have identified the cause of the issue.
In the latest version v.2.0.60-alpha, we have made it possible to specify WASAPI Exclude.
Please give it a try.

Thank you for your continued cooperation.

![image](https://github.com/user-attachments/assets/3455fbd2-44f0-4e68-ac8d-594415c9a36b)

**Comment by JarodMica** [comment]
Looks good, it seems to work without issues.  Appreciate it :)!
